addons/addon-base-raas/packages/base-raas-services/lib/data-source/access-strategy/roles-only/environment-resource-service.js

```diff
@@ -0,0 +1,287 @@
+
+    // For certain calls, the system context is used instead of the request context
+    const systemContext = getSystemRequestContext();
+
+    // groups is an object where the keys are the appRoleArn and the values are arrays
+    // of the studyEntities. Example:
+    // groups = { '<appRoleArn1>': [ <studyEntity>, ... ], '<appRoleArn1>': [ <studyEntity>, ... ]}
+    const groups = _.groupBy(studies, 'appRoleArn');
+    const groupProcessor = async group => {
+      // group is an array of study entities where they all have the same app role arn
+      const appRoleArn = _.get(_.nth(group, 0), 'appRoleArn', '');
+      const lockId = `roles-only-access-app-role-${appRoleArn}`;
+
+      await lockService.tryWriteLockAndRun({ id: lockId }, async () => {
+        // For each study, we ask the filesystem role system service to deallocate the role for us
+        // eslint-disable-next-line no-restricted-syntax
+        for (const study of group) {
+          const fsRoleArn = studyRoles[study.id];
+          // eslint-disable-next-line no-continue
+          if (_.isEmpty(fsRoleArn)) continue;
+          await filesystemRoleService.deallocateRole(
+            systemContext,
+            fsRoleArn,
+            study,
+            environmentScEntity,
+            memberAccountId,
+          );
+
+          delete studyRoles[study.id];
+        }
+      });
+    };
+
+    // We want to process the groups 10 at a time
+    await processInBatches(_.values(groups), 10, groupProcessor);
+
+    // Time to update the study role map in the environment entity
+    await environmentService.updateStudyRoles(systemContext, environmentScEntity.id, studyRoles);
+  }
+
+  /**
+   * Populate the policy document that is going to be used as an inline policy document in the instance profile role
+   * of the compute resource. It is important to keep in mind that other study access strategy might be participating
+   * in this logic.
+   *
+   * @param requestContext The standard request context object
+   * @param policyDoc An instance of the StudyPolicy class
+   * @param studies an array of StudyEntity that are associated with this env. IMPORTANT: each element in the array is
+   * the standard StudyEntity, however, there is one additional attributes added to each of the StudyEntity. This
+   * additional attribute is called 'envPermission', it is an object with the following shape:
+   * { read: true/false, write: true/false }
+   * @param environmentScEntity the EnvironmentSc entity. The entity is expected to have an attributed named 'studyRoles'.
+   * The studyRoles is a map where the key is the study id and the value is the fs role arn
+   *
+   */
+  async provideEnvRolePolicy(requestContext, { policyDoc, studies: allStudies, environmentScEntity }) {
+    // We are only interested in studies with bucket access = 'roles'
+    const studies = _.filter(allStudies, study => study.bucketAccess === 'roles');
```

Comment:
Shouldn't this kind of filter be added in legacy version of environment-resource-service

Reply:
The existing studies and studyService.create() method don't populate this property.   This property is only populated when you register a **bucket** and then you register a study in that bucket.  When you register a **bucket**, the API asks for the access strategy to use when accessing studies in the bucket.   Then, when you register the study (you specify the account and bucket), we  populate the study.bucketAccess value based on the bucket  access strategy.  The legacy studies don't have that.

The best why to tell if a study is "legacy" study is by checking if study.resources is empty/undefined.

Reply:
Got it thanks. Basically study.resouces is not empty.